조기요 점포관리 기능 PR

ksj2083 opened the pull request:
**PR 생성 날짜**
2022-12-05

**이번 PR에서 구현된 기능**
1. 점포 등록
2. 점포 열기/닫기
3. 점포 정보 수정
4. 점포 조회

**특이사항**
1. 사용자 관리 기능이 아직 구현되지 않아 하드코딩된 부분이 있음
2. 점포 정보 수정은 아직 테스트코드 미작성

ksj2083 committed:
점포등록 기능 구현 및 테스트 작성

클라이언트로 부터 CreateStore.Request 를 받아서 점포 등록하는 기능을 구현하였음. user가 존재하지 않거나 status가 "owner" 가 아닐 경우, 예외를 반환함. 예외는 GlobalExceptionHandler 에서 핸들링됨. 현재 user repository가 아직 구현되지 않아서 임시로 하드코딩된 부분이 있음.

ksj2083 committed:
점포열기/닫기 구현 및 테스트 작성

열려있는데 열려고 할 경우, 닫혀있는데 닫으려고 할 경우 AlreadyOpenClosedException 반환

ksj2083 committed:
점포열기/닫기 시간 추가

점포 열기/닫기 시 현재시간 기록하도록 수정

ksj2083 committed:
점포조회 및 테스트 구현

ownerId를 받아서 점포리스트 조회, 현재 ownerId를 통해 DB에서 owner를 가져 올 수 없어서 그 부분 하드코딩 되어있음.

ksj2083 committed:
점포수정 구현 완료

점포 수정 구현, 추후 테스트 작성 예정

KimHyoJin commented on src/main/java/zerobase/demo/owner/controller/OwnerStoreController.java:
```diff
@@ -0,0 +1,48 @@
+
+@RestController
+@RequiredArgsConstructor
+public class OwnerStoreController {
```
사소한거긴하지만 이쪽에 @RequestMapping("/store")로 한번 공통해서 모아주면 좋을 것 같아요. path가 중복되면 어플리케이션 실행이 안되는데요, 추후에 controller class가 많아지면 path가 중복인지 아닌지 찾아보기가 어렵습니다. 그래서 controller마다 controller 네이밍에 맞게 한번 상단에서 묶어주고 시작해주면 보다 수월하게 관리가 가능해요

ksj2083 replied:
수용하여 수정하였습니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/controller/OwnerStoreController.java:
```diff
@@ -0,0 +1,48 @@
+	//점포 열기/닫기
+	@PutMapping("/store/openclose")
+	public void openCloseStore(@RequestBody OpenCloseStore.Request request) {
+		storeService.openCloseStore(request);
```
request를 바로 service layer로 넣는것보단 클래스를 분리하시는걸 추천드려요! 추후 필요한필드 불필요한 필드를 구별하는데 훨씬 도움이 되고, layer별 격리도를 높일 수 있습니다

ksj2083 replied:
수용하여 수정하였습니다. dto 클래스 안에 내부클래스로 Request, Response 클래스 를 만드니 가독성이 좋아졌어요.

KimHyoJin commented on src/main/java/zerobase/demo/owner/dto/StoreInfo.java:
```diff
@@ -0,0 +1,57 @@
+
+	public static List<StoreInfo> fromEntity(List<Store> storeList) {
+
+		List<StoreInfo> list = new ArrayList<>();
```
요렇게 짜셔도 되지만 java8에서 가독성 좋게 짜는 방법이 있습니다
`storeList.stream().map(it -> fromEntity(it)).collect(Collectors.toList());
`를 활용하시면 한줄에 `fromEntity` 함수의 내용을 담을 수 있습니다

ksj2083 replied:
수용하여 수정하였습니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+		//owner 검증
+		if(!optionalUser.isPresent()) throw new NonExistentUserException();
+		if(!Objects.equals(user.getStatus(), "owner")) throw new NotAuthorizedException();
```
https://github.com/dksekfbs72/zogiyo/pull/2 에도 달아드린 리뷰이긴한데 status를 enum으로 관리하면 좀더 좋을 것 같아요!

ksj2083 replied:
수용하여 수정하였습니다. 또한 StoreOpenCloseStatus 또한 enum으로 관리하도록 수정했습니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+	@Override
+	public void updateStore(UpdateStore.Request request) {
+
+		Optional<Store> optionalStore = storeRepository.findById(request.getId());
```
`Store store = storeRepository.findById(request.getId()).orElseThrow(() -> new NonExistentStoreException());
`
로 가독성 좋게 바꿔볼수 있을 것 같아요!

ksj2083 replied:
이부분은 좀 더 고민해 보겠습니다. 모든 예외처리를 람다로 처리할 수는 없어서 통일성을 헤치는 느낌이 들어서요.

ksj2083 replied:
고민해 봤는데 이게 더 가독성이 좋은 것 같아서 수정하겠습니다!

KimHyoJin commented on src/main/java/zerobase/demo/owner/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+		updateStore.setDeliveryTip(request.getDeliveryTip());
+		updateStore.setCommission(request.getCommission());
+
+		updateStore.setUdtDt(LocalDateTime.now());
```
@LastModifiedDate를 쓰면 수정시 업데이트되어야하는 column에 대해서 보다 깔끔하게 처리가 가능합니다! 휴먼에러도 발생시키지 않구요. 아래 블로그 한번 참고해보세요 https://wildeveloperetrain.tistory.com/76

ksj2083 replied:
수용하여 적용했습니다. 생성시간과 수정시간은 추상클래스 BaseEntity 가 가지도록 했습니다. 팀원과 모든 엔티티가 BaseEntity를 상속하도록 구현하는 것으로 협의 하였습니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/controller/OwnerStoreController.java:
```diff
@@ -0,0 +1,48 @@
+
+	//점포 조회
+	@GetMapping("/store")
+	public List<StoreInfo> getStoreByOwnerId(@RequestParam int id) {
```
List형태로 뿌려줘도 되지만 Response용 class를 작성하셔서 이를 이용해서 응답을 내려주셔도 좋을 것 같아요! https://github.com/dksekfbs72/zogiyo/pull/2 같은 스터디원분과 한번 이야기 나눠보시면 좋을 것 같습니다!

ksj2083 replied:
수용하여 처리하였습니다. 관련해서 궁금한게 있는데 exception이 발생했을때, 프론트에게 Exception 자체를 넘기는 것과 Response에 예외코드같은걸 담아서 넘기는 것 중에 어떤게 더 좋은 코드일까요?

KimHyoJin replied:
프론트에게는 Response에 예외코드를 담아서 넘기는 것이 좋습니다! 프론트에게 어떤 exception인지를 exception을 넘겨주는건 큰 의미가 없습니다. exception이 error code보다 좋은 이유는 어디에서 에러가 발생했는지 등의 stacktrace를 볼수 있다는 점입니다. 그런데 프론트는 아예 다른 파트이고 부분이기때문에 해당 정보를 알아도 얻는 이득이 없습니다. 대신 요청에 문제가 있어서 발생한 에러인지, 서버내 문제가 있어서 발생한 에러인지 등을 알필요가 있죠. 따라서 규격화된 상태(response)에 넣어서 응답해주는 것이 좋습니다 :)

ksj2083 replied:
답변 감사합니다. 참고해서 최대한 규격화 해 보려고 시도중입니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	@Override
+	public void updateStore(UpdateStore.Request request) {
```
요청한유저가 점포 수정 권한이 있는 유저인가 대한 체크도 들어가면 좋겠네요!

ksj2083 replied:
이부분 빼먹었네요! 팀원과 머지하고 스프링 시큐리티가 적용되면 로그인한 계정이 대항 점포에 대한 권한이 있는지 확인하는 부분 추가하도록 하겠습니다.

KimHyoJin commented on src/main/java/zerobase/demo/owner/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+		Optional<Store> optionalStore = storeRepository.findById(request.getId());
+
+		if(!optionalStore.isPresent()) throw new NonExistentStoreException();
```
store번호가 exception안에 들어가면 추후 로그만 보고 에러상황을 좀더 쉽게 파악할 수 있을 것 같아요!

ksj2083 replied:
우선 적용하였습니다.  그런데 좀 더 개선이 필요할 것 같아서 다음 pr때 더 개선해 보겠습니다.

KimHyoJin commented on src/test/java/zerobase/demo/owner/OwnerStoreTest.java:
```diff
@@ -0,0 +1,175 @@
+	void createStoreSuccessTest() throws Exception {
+
+		//given
+		Integer ownerId = 1; //임시로 user 테이블에 강제 insert
```
요거 임시로 넣어주셨다는 말씀이신것 같은데, 현재 코드는 integration test로 전체 spring application을 띄워서 테스트하고 계신 방법으로 보여요.
그렇다면 @Before을 이용해서 테스트전 userId를 하나 만들어서 넣어주고, 이후 @After를 통해 user를 삭제해줌으로서 데이터베이스를 깔끔하게 관리하고, 어떠한 사전작업 없이도, 예를 들어 데이터베이스를 새로 만들거나 옮겨졌을때도 테스트는 정상 동작할 수 있도록 해주는 것이 좋을 것 같습니다. 특히 테스트시에는 데이터베이스를 h2등의 휘발성 테이터베이스를 많이 사용합니다. (https://jungguji.github.io/2020/07/05/H2-DB%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-Repository-Test/)
이런경우 테이블에 강제로 넣어주기가 어렵기 때문에 위 방법을 사용하시는걸 추천드려요!

참고 하실만한 블로그도 같이 올려드립니다 https://blusky10.tistory.com/508

KimHyoJin reviewed (approved):
코드가 굉장히 깔끔하시네요!! 수고 많으셨습니다. 간단히 몇가지 코멘트 드려보았어요. 필요하시다고 생각하시는 부분들을 적용해보시면 좋을 것 같습니다 :)

ksj2083 committed:
Layer 격리도 향상 및 코드 가독성 개선

1. Controller에 @RequestMapping 사용하여 가독성 개선
2. BaseEntity 추가, 앞으로 모든 엔티티는 BaseEntity를 상속할 예정
3. Controller 에서 service로 param을 내릴 때, request를 바로 내리지 않고 dto로 변환하여 내리도록 구조를 변경하여 layer별 격리도를 높혔음. 또한 Requst는 dto 클래스의 내부클래스로 구현하여 가독성을 높혔음
4. Controller 에서 client로 결과를 전송 할 때, 단순히 List만 보내지 않고 Response 클래스에 담아서 보내도록 수정 (OwnerStoreController.getStoreByOwnerId(int) 메소드에 적용)
5. StoreInfo.fromEntity(List<Store>) 함수, 람다식을 사용해여 가독성 개선

ksj2083 committed:
UserStatus, StoreOpenCloseStatus 을 enum 으로 관리하도록 수정

유저 등급과 점포오픈 여부를 enum 타입으로 관리하도록 하여 가독성을 높혔음

ksj2083 committed:
regDt, udtDt에 @LastModifiedDate 적용

등록시간과 수정시간 자동 입력되도록 하여 가독성을 올리고 휴먼에러의 여지를 없앴음

ksj2083 committed:
createStore 시 오픈/클로즈 타임 기본값 입력하도록 수정

openCloseDt기본값은 점포 생성시간임. 참고로 openClose 기본값은 close임

ksj2083 committed:
Exception 관련 버그 수정 / 불합리 개선 / 예외 세부정보 추가

1. GloblaExceptionHandler에 NonExistentUserException과 NotAuthorizedException 처리하는 부분이 없던 버그 수정
2. GlobalExceptionHandler 에서 핸들링하고 있는 예외들 처리할때 파라미터를 exception 자체를 받아서 리턴하는 것으로 변경
   - 기존에는 exception 을 서비스단에서도 생성하고 GlobalException에서도 생성하는 불합리가 있었음
3. NonExistentUserException, NonExistentStoreException 예외 발생 시 요청했던 유저번호, 점포번호도 함께 return 하도록 수정

ksj2083 committed:
1차 합의
